File: 백준/Silver/3187. 양치기 꿍/양치기 꿍.js
const fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().split('\n');

//bfs 함수
function bfs(i,j){
    queue.push([i,j]);
    visited[i][j] = true
    let v_cnt = 0;
    let k_cnt = 0;

    while(queue.length){
        let [x,y] = queue.shift();
        if(board[x][y] === 'v'){
            v_cnt += 1
        } else if(board[x][y] === 'k') {
            k_cnt += 1
        } 
        for(var k=0; k<4; k++){
            let nx = x+dx[k]
            let ny = y+dy[k]
            if(0<=nx && nx<R && 0<=ny && ny<C && !visited[nx][ny]){
                if(board[nx][ny] !== '#'){
                    queue.push([nx,ny]);
                    visited[nx][ny] = true
                }
            }
        }
    }

    k_cnt > v_cnt ? v_cnt = 0 : k_cnt = 0;

    return [v_cnt,k_cnt]
}

const [R,C] = input.shift().split(' ').map(Number);
const board = input.map((v) => v.split(''));

let visited = Array.from(Array(R),() => Array(C).fill(false));
let dx = [-1,1,0,0];
let dy = [0,0,-1,1];

let queue = [];

// 늑대와 양의 수
let v = 0;
let k = 0;

// 울타리가 아니면 bfs돌리기
for(var i=0; i<R; i++){
    for(var j=0; j<C; j++){
        if(board[i][j] !== '#' && !visited[i][j]) {
            let [v_cnt,k_cnt] = bfs(i,j)
            v += v_cnt
            k += k_cnt
        }
    }
}

console.log(k+' '+v);

File: 백준/Silver/3187. 양치기 꿍/양치기 꿍.py
# 울타리 영역 안의 양들의 숫자 > 늑대의 숫자: 늑대가 전부 먹힘
# 양들의 숫자 <= 늑대의 숫자 : 양이 전부 먹힘 
# 몇마리의 양과 늑대가 살아남을 수 있을지 return 
# 동서남북으로만 이동 가능
# v: 늑대, k:양

import sys
input = sys.stdin.readline
from collections import deque

def bfs(i,j):
    queue.append([i,j])
    visited[i][j] = True
    v_cnt,k_cnt = 0,0

    while queue:
        x,y = queue.popleft()
        if board[x][y] == 'v':
            v_cnt += 1
        elif board[x][y] == 'k':
            k_cnt += 1
        # '.'
        for k in range(4):
            nx = x+dx[k]
            ny = y+dy[k]
            if 0<=nx<R and 0<=ny<C and not visited[nx][ny]:
                if board[nx][ny] != '#':
                    queue.append([nx,ny])
                    visited[nx][ny] = True
    
    if v_cnt >= k_cnt:
        k_cnt = 0
    elif v_cnt < k_cnt:
        v_cnt = 0

    return v_cnt,k_cnt

R,C = map(int,input().split())
board = [list(input().rstrip()) for _ in range(R)]

dx = [-1,1,0,0]
dy = [0,0,-1,1]
visited = [[False]*C for _ in range(R)]
queue = deque([])
v,k = 0,0

for i in range(R):
    for j in range(C):
        if board[i][j] != '#' and not visited[i][j]:
            v_cnt,k_cnt = bfs(i,j)
            v+=v_cnt
            k+=k_cnt

print(k,v)